src/xwayland/xwm/surface.rs

```diff
@@ -0,0 +1,878 @@
+        self == &window.window
+    }
+}
+
+impl X11Relatable for WlSurface {
+    fn is_window(&self, window: &X11Surface) -> bool {
+        let state = window.state.lock().unwrap();
+        state
+            .wl_surface
+            .as_ref()
+            .map(|surface| surface == self)
+            .unwrap_or(false)
+    }
+}
+
+impl X11Relatable for Id {
+    fn is_window(&self, window: &X11Surface) -> bool {
+        let state = window.state.lock().unwrap();
+        state
+            .wl_surface
+            .as_ref()
+            .map(Id::from_wayland_resource)
+            .map(|id| &id == self)
+            .unwrap_or(false)
+    }
+}
+
+impl IsAlive for X11Surface {
+    fn alive(&self) -> bool {
+        self.state.lock().unwrap().alive
+    }
+}
+
+impl<D: SeatHandler + 'static> KeyboardTarget<D> for X11Surface {
+    fn enter(&self, seat: &Seat<D>, data: &mut D, keys: Vec<KeysymHandle<'_>>, serial: Serial) {
+        match self.input_mode() {
+            InputMode::None => return,
+            InputMode::Passive => {
+                if let Some(conn) = self.conn.upgrade() {
+                    if let Err(err) = conn.set_input_focus(InputFocus::NONE, self.window, x11rb::CURRENT_TIME)
+                    {
+                        slog::warn!(
+                            self.log,
+                            "Unable to set focus for X11Surface ({:?}): {}",
+                            self.window,
+                            err
+                        );
+                    }
+                    let _ = conn.flush();
+                }
+            }
+            InputMode::LocallyActive => {
+                if let Some(conn) = self.conn.upgrade() {
+                    let event = ClientMessageEvent::new(
+                        32,
+                        self.window,
+                        self.atoms.WM_PROTOCOLS,
+                        [self.atoms.WM_TAKE_FOCUS, x11rb::CURRENT_TIME, 0, 0, 0],
+                    );
+                    if let Err(err) = conn.send_event(false, self.window, EventMask::NO_EVENT, event) {
+                        slog::warn!(
+                            self.log,
+                            "Unable to send take focus event for X11Surface ({:?}): {}",
+                            self.window,
+                            err
+                        );
+                    }
+                    let _ = conn.flush();
+                }
+            }
+            _ => {}
+        };
+        if let Some(surface) = self.state.lock().unwrap().wl_surface.as_ref() {
+            KeyboardTarget::enter(surface, seat, data, keys, serial);
+        }
+    }
+
+    fn leave(&self, seat: &Seat<D>, data: &mut D, serial: Serial) {
+        if self.input_mode() == InputMode::None {
+            return;
+        } else if let Some(conn) = self.conn.upgrade() {
+            if let Err(err) = conn.set_input_focus(InputFocus::NONE, x11rb::NONE, x11rb::CURRENT_TIME) {
```

Comment:
Hm... :thinking: 

Xwayland really is a special case. In a "normal WM" this would be a really bad idea. But here it might actually be fine to send all input to `/dev/null`...

Reply:
Yes, a normal X11 Server would likely always have something focused. But with wayland surfaces being in the picture this seems to be the correct solution and so far this hasn't caused any issue. I'll just remove this line and keep the focus on the last X11 window, if any clients actually start misbehaving because of it.